FirebaseMLModelDownloader/Tests/Unit/ModelDownloaderUnitTests.swift

```diff
@@ -148,3 +144,376 @@ final class ModelDownloaderUnitTests: XCTestCase {
+                               isDirectory: true)
+    let tempFileURL = tempDirectoryURL.appendingPathComponent("fake-model-file.tmp")
+    let tempData: Data = "fakeModelData".data(using: .utf8)!
+    try? tempData.write(to: tempFileURL)
+
+    // Expect file downloader to be called.
+    let downloader = MockModelFileDownloader()
+    let fileDownloaderExpectation = expectation(description: "file downloader")
+    downloader.downloadFileHandler = { url in
+      fileDownloaderExpectation.fulfill()
+      XCTAssertEqual(url, fileDownloadURL)
+    }
+
+    // Create task and set up progress and completion expectations.
+    let progressExpectation = expectation(description: "progress handler")
+    progressExpectation.expectedFulfillmentCount = 2
+    let completionExpectation = expectation(description: "completion handler")
+
+    let modelDownloadTask = ModelDownloadTask(remoteModelInfo: fakeRemoteModelInfo,
+                                              appName: "fakeAppName",
+                                              defaults: .createTestInstance(testName: #function),
+                                              downloader: downloader,
+                                              modelInfoRetriever: modelInfoRetriever,
+                                              progressHandler: {
+                                                progress in
+                                                progressExpectation.fulfill()
+                                                XCTAssertEqual(progress, 0.4)
+                                              }) { result in
+      completionExpectation.fulfill()
+      switch result {
+      case let .success(model):
+        XCTAssertEqual(model.name, self.fakeModelName)
+        XCTAssertEqual(model.size, self.fakeModelSize)
+        XCTAssertEqual(model.hash, self.fakeModelHash)
+        XCTAssertTrue(ModelFileManager.isFileReachable(at: URL(string: model.path)!))
+      case let .failure(error):
+        XCTFail("Error - \(error)")
+      }
+    }
+
+    modelDownloadTask.download()
+
+    // Wait for downloader to be called.
+    wait(for: [fileDownloaderExpectation], timeout: 0.1)
+
+    // Call downloader progress handler and wait for task progress handler to be called.
+    downloader.progressHandler?(100, 250)
+    // Call a second time as we expect the task progress handler to be called twice.
+    downloader.progressHandler?(100, 250)
+    wait(for: [progressExpectation], timeout: 0.5)
+
+    // Call download completion and wait for task completion.
+    downloader
+      .completion?(.success(FileDownloaderResponse(urlResponse: fakeResponse,
+                                                   fileURL: tempFileURL)))
+
+    wait(for: [completionExpectation], timeout: 0.5)
+
+    // Cleanup temp file.
+    try? ModelFileManager.removeFile(at: tempFileURL)
+  }
+
+  // TODO: Update similarly to `testModelDownloadWith200`
```

Comment:
Would you like to address the todo in the next PR?

Reply:
Ah sorry I missed this. I've added one test that checks for successful retry (i.e., task has expired url at first but retrieves a subsequent valid url). In order to test failed retry, I think it's best to first refactor `maybeRetryDownload()` to be in `ModelDownloader` (upcoming PR).